rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,427 @@
+    usually required to invoke the training step multiple times. It's doable by
+    writing a loop in application code to invoke the training subgraph multiple
+    times. It can also be implemented by using a `While` op to iterate over the
+    training data.
+
+The detailed design of supporting control flow in TensorFlow Lite is at
+[This RFC](https://github.com/tensorflow/community/pull/83).
+
+### Tensor List
+
+Tensor List is not required for basic training. However it's important for
+training models with control flow loops.
+
+Whenever there are loops (e.g. While) in the inference graph, Tensor List is
+usually required to train the model. Even if Tensor List ops aren't in the
+original inference graph, when constructing the training graph, Tensor List ops
+are often added automatically to memorize intermediate values, for computing
+gradients over the loop.
+
+Tensor Lists can also be used without loops but this rarely happens. In this
+case, Tensor List ops are required for inference and training.
+
+### Saving Format
+
+We don't need a saving format to make training work. The newly trained weights
+can be used in TensorFlow Lite directly for inference. The weights data are
+stored in (variable) tensors, and users can read and write these data through
+interpreter API. This already enables saving and loading trainable weights from
+the application code. On top of this, we can define a saving format to store
+trainable weights.
+
+**(Open question)** There are a few potential options:
+
+*   Option 0: Do nothing
+*   Option 1: Implement TensorFlow Lite's own variable saving file format. This
+    can be a simple binary format that stores variable ID -> data mapping.
+*   Option 2: Write the variable value back to TensorFlow Lite model.
+    *   Pros: Only having a single file & single file format.
+    *   Cons: Have to ship FlatBuffer writer in the binary. Have to write the
+        entire model (which may contain other huge frozen constants) back to
+        storage.
+*   Option 3: Use TensorFlow checkpoint format
+    *   3A: Implement checkpoint reading / writing functionality in TensorFlow
+        Lite
+    *   3B: Reading / writing through Flex runtime (so it only works when Flex
+        runtime is used)
+
+### Feeding data with tf.Example / tf.DataSet
+
+`tf.Example` and `tf.DataSet` are technically not training-specific
+requirements. These features can also be used in inference.
+
+The basic training can work by feeding the raw data (e.g. float values) into
+TensorFlow Lite interpreter. On top of this, we can support tf.Example and
+tf.DataSet to make this easier to use (likely via Flex runtime). There may be a
+performance drawback using these features.
+
+### Selective Registration
+
+Initially we will need to rely on Flex runtime to execute some of the fused
+gradient ops. This requires to add the gradient op kernels to Flex runtime, and
+it will bloat up the Flex binary size.
```

Comment:
How much bloat are we talking? What percent increase is expected over the existing binary size?

Reply:
As far as I remember, TFLite interpreter (with no kernel) is around 80K, and it's around ~700K with all the builtin ops. The Flex runtime is at least 2MB. This are binary library size for one architecture (e.g. arm64). 

These numbers may be a few months old and TFLite is changing fast. I can find out the newest number if necessary.